### PR TITLE
#481 accept date and time as string in ISO 8601 format for ChronoLocalDateTime [WIP]

### PR DESCRIPTION
**Pull request by rkiselev**
Hi, @robstoll, could you please check, am I on the right way? I'm stuck on 'add the string version to ChronoLocalDateTimeAssertionsSpec:'
First of all, I assume that ChronoLocalDateTimeAssertionsSpec is ChronoLocalDateTimeAssertionSpec.
Second, I'm new at Kotlin, so, could you provide some more information on how to implement this part.
______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

**Comment by robstoll** [comment]
@rkiselev already some feedback. I'll give you some hints about the spec and try to answer your other questions tomorrow.

**Comment by robstoll** [comment]
> First of all, I assume that ChronoLocalDateTimeAssertionsSpec is ChronoLocalDateTimeAssertionSpec.

Yes, ChronoLocalDateTimeAssertionsSpec is indeed ChronoLocalDateTimeAssertionsSpec :laughing: 
It was also a copy&paste error on my part, it should not have been ChronoLocalDateAssertionsSpec but ChronoLocalDateTimeAssertionsSpec correct.

> Second, I'm new at Kotlin, so, could you provide some more information on how to implement this part.

I'll comment directly on the code

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateTimeAssertionSpec.kt [hunk 1]
```diff
@@ -20,6 +21,14 @@ class ChronoLocalDateTimeAssertionSpec : Spek({
         fun1(Expect<ChronoLocalDateTime<*>>::isEqual)
     )
 
+    object StringSpec : ch.tutteli.atrium.specs.integration.ChronoLocalDateTimeAssertionSpec(
+        fun1(Expect<ChronoLocalDateTime<*>>::isBefore),
```
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateTimeAssertionSpec.kt:25]: The goal is that we use not the function which expects a ChronoLocalDateTime but your function which expects a String. Hence create a function in the companion object with signature:
```
fun isBefore(expect: Expect<ChronoLocalDateTime<*>>, expected: ChronoLocalDateTime<*>) 
```
Which converts `expected` into a string an passes it to your string function. 
Same same but different for the other methods
```

**Comment by rkiselev** [comment]
thx for your answers, I will continue work on weekends.

File: logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt [hunk 2]
```diff
@@ -14,29 +16,72 @@ class DefaultChronoLocalDateTimeAssertions : ChronoLocalDateTimeAssertions {
         expected: ChronoLocalDateTime<*>
     ): Assertion = container.createDescriptiveAssertion(IS_BEFORE, expected) { it.isBefore(expected) }
 
+    override fun <T : ChronoLocalDateTime<out ChronoLocalDate>> isBefore(
+        container: AssertionContainer<T>,
+        expected: String
+    ): Assertion = container.createDescriptiveAssertion(IS_BEFORE, expected) {
+        it.isBefore(stringToLocalDateTime(expected))
+    }
```
Comment (robstoll) [inline logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt:29]: please delegate to the overload which expects `ChronoLocalDateTime<*>`
```suggestion
    ): Assertion = container.isBefore(stringToLocalDateTime(expected))
```

File: logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt [hunk 3]
```diff
@@ -14,29 +16,72 @@ class DefaultChronoLocalDateTimeAssertions : ChronoLocalDateTimeAssertions {
...
     ): Assertion = container.createDescriptiveAssertion(IS_BEFORE_OR_EQUAL, expected) {
         it.isBefore(expected) || it.isEqual(expected)
     }
 
+    override fun <T : ChronoLocalDateTime<out ChronoLocalDate>> isBeforeOrEqual(
+        container: AssertionContainer<T>,
+        expected: String
+    ): Assertion = container.createDescriptiveAssertion(IS_BEFORE_OR_EQUAL, expected) {
+        it.isBefore(stringToLocalDateTime(expected)) || it.isEqual(stringToLocalDateTime(expected))
+    }
```
Comment (robstoll) [inline logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt:43]:
```suggestion
    ): Assertion = container.isBeforeOrEqual(stringToLocalDateTime(expected))
```

File: logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt [hunk 4]
```diff
@@ -14,29 +16,72 @@ class DefaultChronoLocalDateTimeAssertions : ChronoLocalDateTimeAssertions {
...
         container: AssertionContainer<T>,
         expected: ChronoLocalDateTime<*>
     ): Assertion = container.createDescriptiveAssertion(IS_AFTER, expected) { it.isAfter(expected) }
 
+    override fun <T : ChronoLocalDateTime<out ChronoLocalDate>> isAfter(
+        container: AssertionContainer<T>,
+        expected: String
+    ): Assertion = container.createDescriptiveAssertion(IS_AFTER, expected) {
+        it.isAfter(stringToLocalDateTime(expected))
+    }
```
Comment (robstoll) [inline logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt:55]:
```suggestion
    ): Assertion = isAfter(stringToLocalDateTime(expected))
```

File: logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt [hunk 5]
```diff
@@ -14,29 +16,72 @@ class DefaultChronoLocalDateTimeAssertions : ChronoLocalDateTimeAssertions {
...
     ): Assertion = container.createDescriptiveAssertion(IS_AFTER_OR_EQUAL, expected) {
         it.isAfter(expected) || it.isEqual(expected)
     }
 
+    override fun <T : ChronoLocalDateTime<out ChronoLocalDate>> isAfterOrEqual(
+        container: AssertionContainer<T>,
+        expected: String
+    ): Assertion = container.createDescriptiveAssertion(IS_AFTER_OR_EQUAL, expected) {
+        it.isAfter(stringToLocalDateTime(expected)) || it.isEqual(stringToLocalDateTime(expected))
+    }
```
Comment (robstoll) [inline logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt:69]:
```suggestion
    ): Assertion = container.isAfterOrEqual(stringToLocalDateTime(expected))
```

File: logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt [hunk 6]
```diff
@@ -14,29 +16,72 @@ class DefaultChronoLocalDateTimeAssertions : ChronoLocalDateTimeAssertions {
...
     ): Assertion = container.createDescriptiveAssertion(IS_EQUAL_TO, expected) {
         it.isEqual(expected)
     }
+
+    override fun <T : ChronoLocalDateTime<out ChronoLocalDate>> isEqual(
+        container: AssertionContainer<T>,
+        expected: String
+    ): Assertion = container.createDescriptiveAssertion(IS_EQUAL_TO, expected) {
+        it.isEqual(stringToLocalDateTime(expected))
+    }
```
Comment (robstoll) [inline logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt:83]:
```suggestion
    ): Assertion = container.isEqual(stringToLocalDateTime(expected))
```

**Comment by robstoll** [comment]
@rkiselev blimey, now I was a bit too fast with merging your PR, I thought everything was done after my fix. You can pick up the work from branch `#481-ChronoLocalDateTime-as-string`:
```
git remote add upstream https://github.com/robstoll/atrium.git
git fetch upstream
git checkout -b upstream/#481-ChronoLocalDateTime-as-string
```

I have added an example to `ChronoLocalDateTimeAssertionSpec` how to refer to the companion function (see TODOs)

**Comment by rkiselev** [comment]
Hi, @robstoll, I get an error, looks like there is not #481-ChronoLocalDateTime-as-string branch

**Comment by robstoll** [comment]
@rkiselev oho, seems like I deleted it inadvertently. I'll should have it locally but on a different machine, I'll look at it on wednesday. Sorry for the trouble.

**Comment by rkiselev** [comment]
ok, will wait.

**Comment by robstoll** [comment]
@rkiselev I am afraid but I don't have access to the machine today. You could copy & paste the following files if you want
(based on https://github.com/robstoll/atrium/commit/def1fc9a99e48101b034623a1b44262549469f0e):
- https://raw.githubusercontent.com/robstoll/atrium/def1fc9a99e48101b034623a1b44262549469f0e/apis/fluent-en_GB/atrium-api-fluent-en_GB-jvm/src/test/kotlin/ch/tutteli/atrium/api/fluent/en_GB/ChronoLocalDateTimeAssertionSpec.kt
- https://raw.githubusercontent.com/robstoll/atrium/def1fc9a99e48101b034623a1b44262549469f0e/logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/ChronoLocalDateTimeAssertions.kt
- https://raw.githubusercontent.com/robstoll/atrium/def1fc9a99e48101b034623a1b44262549469f0e/logic/atrium-logic-jvm/src/main/kotlin/ch/tutteli/atrium/logic/impl/DefaultChronoLocalDateTimeAssertions.kt
- https://raw.githubusercontent.com/robstoll/atrium/def1fc9a99e48101b034623a1b44262549469f0e/misc/deprecated/domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/chronoLocalDateTimeAssertions.kt
- https://raw.githubusercontent.com/robstoll/atrium/def1fc9a99e48101b034623a1b44262549469f0e/misc/deprecated/domain/robstoll/atrium-domain-robstoll-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/creating/ChronoLocalDateTimeAssertionsImpl.kt

run `gradlew generateLogic` afterwards

**Comment by robstoll** [comment]
@rkiselev I have pushed the branch again

**Pull request merged**